oak_crypto/src/noise_nk/rustcrypto.rs

```diff
@@ -0,0 +1,212 @@
+    elliptic_curve::{
+        ops::{Mul, MulByGenerator},
+        sec1::{FromEncodedPoint, ToEncodedPoint},
+    },
+    Field, PrimeField,
+};
+use sha2::Digest;
+
+pub fn rand_bytes(_output: &mut [u8]) {
+    panic!("unimplemented");
+}
+
+/// Perform the HKDF operation from https://datatracker.ietf.org/doc/html/rfc5869
+pub fn hkdf_sha256(ikm: &[u8], salt: &[u8], info: &[u8], output: &mut [u8]) -> Result<(), ()> {
+    hkdf::Hkdf::<sha2::Sha256>::new(Some(salt), ikm).expand(info, output).map_err(|_| ())
+}
+
+pub fn aes_256_gcm_seal_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    plaintext: &mut Vec<u8>,
+) {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .encrypt_in_place(nonce.into(), aad, plaintext)
+        .unwrap();
+}
+
+pub fn aes_256_gcm_open_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    mut ciphertext: Vec<u8>,
+) -> Result<Vec<u8>, ()> {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .decrypt_in_place(nonce.into(), aad, &mut ciphertext)
+        .map_err(|_| ())?;
+    Ok(ciphertext)
+}
+
+pub fn sha256(input: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input);
+    ctx.finalize().into()
+}
+
+/// Compute the SHA-256 hash of the concatenation of two inputs.
+pub fn sha256_two_part(input1: &[u8], input2: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input1);
+    ctx.update(input2);
+    ctx.finalize().into()
+}
+
+pub struct P256Scalar {
+    v: p256::Scalar,
+}
+
+impl P256Scalar {
+    pub fn generate() -> P256Scalar {
+        let mut ret = [0u8; P256_SCALAR_LENGTH];
+        // Warning: not very random.
+        ret[0] = 1;
+        P256Scalar { v: p256::Scalar::from_repr(ret.into()).unwrap() }
+    }
+
+    pub fn compute_public_key(&self) -> [u8; P256_X962_LENGTH] {
+        p256::ProjectivePoint::mul_by_generator(&self.v)
+            .to_encoded_point(false)
+            .as_bytes()
+            .try_into()
+            .unwrap()
+    }
+
+    pub fn bytes(&self) -> [u8; P256_SCALAR_LENGTH] {
+        self.v.to_repr().as_slice().try_into().unwrap()
+    }
+}
+
+impl TryFrom<&[u8]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8]) -> Result<Self, ()> {
+        let array: [u8; P256_SCALAR_LENGTH] = bytes.try_into().map_err(|_| ())?;
+        (&array).try_into()
+    }
+}
+
+impl TryFrom<&[u8; P256_SCALAR_LENGTH]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8; P256_SCALAR_LENGTH]) -> Result<Self, ()> {
+        let scalar = p256::Scalar::from_repr((*bytes).into());
+        if !bool::from(scalar.is_some()) {
+            return Err(());
+        }
+        let scalar = scalar.unwrap();
+        if scalar.is_zero_vartime() {
+            return Err(());
+        }
+        Ok(P256Scalar { v: scalar })
+    }
+}
+
+pub fn p256_scalar_mult(
```

Comment:
Shouldn't we be using https://docs.rs/elliptic-curve/latest/elliptic_curve/ecdh/fn.diffie_hellman.html or something already made that can do a DH exchange directly?

Reply:
Maybe.  We should revisit this in a follow-on PR where we either clean up the crypto_wrapper API and add BoringSSL bindings, or flatten it into noise.rs and mod.rs.

I _think_ this is done with scalar point multiplication because that was available in all 3 crypto libraries supported in Chromium's version of this code.

Reply:
Let's do this in a follow-on PR when we either squash crypto_wrapper.rs into noise.rs and mod.rs, or clean up the APIs.  I've seen scalar multiplication more often than a DH function, so it might be simpler to use the more common one in a wrapper supporting multiple back-ends, but we can revisit this in that PR.